### ISSUE TITLE
Compiler does not resolve types as expected in some cases

### ISSUE THREAD

**Issue by YoshihiroUnno** [swiftlang/swift#65533]
**Description**

The Swift compiler does not resolve types as expected in some cases.
I can't exactly describe the conditions under which the problem occurs.

**Steps to reproduce**

Given the following protocols(`P1`, `P2`),

```swift
protocol P1<Item> {
    associatedtype Item
    
    var item: Item { get }
}

protocol P2 {
    associatedtype Item
    
    var anyP1: any P1<Item> { get }
}

extension P2 {
    var anyP1: any P1<Item> {
        Swift.fatalError()
    }
}
```

Among the declarations of the following structures (`S1` to `S5`) conforming to protocol `P2`, especially in `S3` and `S4`,
The type of property `anyP1.item` is not resolved as expected.

```swift
struct S1<G1>: P2 {
    typealias Item = G1
    
    func f1() {
        let item = anyP1.item   /// The compiler resolved the type of `item` as `G1` as expected.
    }
}

class C {}

struct S2: P2 {
    typealias Item = C
    
    func f2() {
        let item = anyP1.item   /// The compiler resolved the type of `item` as `C` as expected.
    }
}

struct S3<G3: C>: P2 {
    typealias Item = G3
    
    func f3() {
        let item = anyP1.item   /// The compiler resolved the type of `item` as `Any` ( expected `G3` ).
    }
}

protocol P {}

struct S4: P2 {
    typealias Item = any P
    
    func f4() {
        let item = anyP1.item   /// The compiler resolved the type of `item` as `P` ( expected `any P` ).
    }
}

struct S5<G5: P>: P2 {
    typealias Item = G5
    
    func f5() {
        let item = anyP1.item   /// The compiler resolved the type of `item` as `G5` as expected.
    }
}
```

**Environment**
- Swift compiler version info: swift-driver version: 1.75.2 Apple Swift version 5.8 (swiftlang-5.8.0.124.2 clang-1403.0.22.11.100)
- Xcode version info: Xcode 14.3 Build version 14E222b
- Deployment target:

**Comment by YoshihiroUnno** [comment]
Thank you.

> > /// The compiler resolved the type of `item` as `P` ( expected `any P` ).
> 
> That’s `any P`, we’re just printing it wrong somewhere. Where is it showing up this way?

I found it in Xcode's Help Inspector. That's the only basis for my opinion.

![スクリーンショット 2023-05-05 1 29 21](https://user-images.githubusercontent.com/120902417/236269492-c68c5e24-ef30-4cd7-9fd3-d94647eee87f.png)
![スクリーンショット 2023-05-05 1 27 45](https://user-images.githubusercontent.com/120902417/236269678-ca3aa096-a829-4964-a7da-5155470692a4.png)

**Comment by AnthonyLatsis** [comment]
@hborla We're not supposed to suppress `any` when printing for Quick Help, are we? `any` printing is even disabled by default, which doesn't seem right.

**Comment by hborla** [comment]
> We're not supposed to suppress any when printing for Quick Help, are we? any printing is even disabled by default, which doesn't seem right.

Yeah, `any` should be shown everywhere, even if it's implicit. I only added the printing flag initially because there were a _ton_ of SIL tests to update, but I fixed those a while back. We can probably remove the printing flag now and fix whatever is left!

**Comment by AnthonyLatsis** [comment]
>  We can probably remove the printing flag now and fix whatever is left!

#65731

**Comment by YoshihiroUnno** [comment]
I am very happy that you responded to the problem so quickly.

**Comment by AnthonyLatsis** [comment]
> I am very happy that you responded to the problem so quickly.

This is the way.
